lib/network/service/event/tunnel.ex

```diff
@@ -0,0 +1,13 @@
+defmodule Helix.Network.Service.Event.Tunnel do
+
+  alias Helix.Network.Model.ConnectionClosedEvent
+  alias Helix.Network.Controller.Tunnel
+
+  def connection_closed(e = %ConnectionClosedEvent{}) do
+    tunnel = Tunnel.fetch(e.tunnel_id)
+
+    if Enum.empty?(Tunnel.get_connections(tunnel)) do
+      Tunnel.delete(tunnel)
+    end
```

Comment:
About 1, it could be fixed by having a running genserver called "tunnel janitor", it could hold the tunnels to be removed on its state and use `erlang:send_after` to delay the delete (and naturally would remove the tunnel deletion from it's queue if an `connection started` event is emited in the meantime

About 2, we technically could do that check with a single index-only cheap query without even fetching the tunnel.

Again, i'll leave that as a debt to the future